---
title: "Tutorial"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Tutorial}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library("RVcov")
library("data.table")
library("stringr")
```

## Performing Rare-Variant Aggregation Testing Given Annotation File

In this vignette we will walk through an example of using the package on summary 
statistics generated using 503 Europeans from 1000G and a reference panel 
of 661 Africans from 1000G. A detailed description of how the data was generated 
can be found in the README file in the extdata/ folder. 
To perform aggregation testing, we need to pass
at a minimum the following parameters to the function *agg_test()*:

**score_stat_file**: The file name and path of the score statistic file.

**vcf_file**: The file name and path of the VCF to be used as the reference panel.

**chr**: The chromosome used in the analysis. Note that this needs to be in the 
same format as in the VCF, which in this case is chr2.

At least one of **burden**, **wburden**, or **SKAT** to be set to TRUE. By default 
**burden** is TRUE and the other 2 are FALSE. 

**two_stage**: Whether or not to perform the two-stage approach. Defaults to FALSE. 
If TRUE, we need to also specify **two_stage_threshold**, which is by default set
to 3. 

Either **group_file**: Vector of strings containing the file paths and names of
the group files to perform the rare-variant aggregation tests with. Or: at least 
one of **pLOF**, **pLOF_narrowMissense**, or **pLOF_broadMissense** to be set to 
TRUE. If the latter is used, we also need to specify **gene**: the name of the 
gene we want to use, and **anno_file** or **anno**: 
The former being the file name and path to a VEP annotation 
file, the latter being an annotation file in VEP output format that has already 
been read into R. 

We may also want to specify **gene_start** and **gene_end**: The start and end 
points of the gene to perform the aggregation testing on. 

For the purposes of this vignette, the score statistic file is:

```{r}
score_stat_file <- "vignetteData/NEB.european.Phenotype.singlevar.score.txt.gz"
```

In this example we are going to use the African subset of the 1000G whole 
genome sequencing data as our reference panel. Although we don't recommend using 
mismatched reference panels like this in general, there was not a large difference
in performance when doing so in our paper. Since we used the European subset of 
1000G to generate the summary statistics, there were no publicly available European
samples remaining to use as the reference panel. The VCF containing the African
subset of the 1000G whole genome sequencing data for the *NEB* gene is available 
with this package, and the file is:

```{r}
vcf_file <- "vignetteData/chr2.1000G.african_unrel.NEB.vcf.gz"
```


Now note that the phenotype was generated using the *NEB* gene on chromosome
2, which has gene boundaries 151485339-151734476. So to call *agg_test()* we can
set **chr** to 2, **gene_start** to 151485339, and **gene_end** to 151734476. 

```{r}
chr <- "chr2"
gene_start <- 151485339
gene_end <- 151734476
```

In this example we will also perform each of the simple and weighted burden test 
as well as SKAT: 

```{r}
burden <- wburden <- SKAT <- TRUE
```

Let us also perform the two stage approach with the default threshold of 3:

```{r}
two_stage <- TRUE
two_stage_threshold <- 3
```


Also included in this package is an annotation file distributed from 1000G. 
So in this first section, we will supply our own annotation file and none of our
own group files. In the second section we will supply our own group files. Note 
that it is not in VEP format, and passing this annotation file directly to *agg_test()*
will either cause an error or lead to unexpected behavior. Hence we will read in 
the annotation file in this vignette, and change the columns so that they are in 
VEP format. After doing so, we will use each of the 3 possible annotation mask
definitions (pLOF, pLOF + missense(narrow), pLOF + missense(broad)) to perform 
aggregation testing. 

```{r}
anno_file <- "vignetteData/chr2.1000G.NEB.rare_variants.annotation.txt"
df_anno <- fread(anno_file, header = TRUE, data.table = FALSE)
```

Note that there are many columns in the annotation file: 

```{r}
dim(df_anno)
```

However this R package assumes only the following columns are present in an
annotation file: 

**PICK**: A column that is 1 if the annotation is to be used or not. We will set this 
to 1 for all variants since this filtering step should have already been done by 1000G. 

**Uploaded_variation**: A column containing the variant in the format CHROM:POS:REF:ALT.

**Consequence**: The predicted consequence of the variant. If there are multiple
consequences, they must be separated by commas. Note that the consequences in the 
1000G data are separated by '&', and must be changed first.

**LRT_pred**: The LRT deleteriousness prediction for the variant. Note that the 
1000G annotation file does not contain LRT predictions, so we will set this to "D"
for all variants so that no variant is excluded from our masks due to this prediction.

**MutationTaster_pred**: MutationTaster deleteriousness prediction. 

**Polyphen2_HDIV_pred**: Polyphen2_HDIV deleteriousness prediction. 

**Polyphen2_HVAR_pred**: Polyphen2_HVAR deleteriousness prediction.

**SIFT4G_pred**: SIFT4G deleteriousness prediction. The 1000G annotations have 
only SIFT predictions so we will use those in place of SIFT4G predictions. 

**SYMBOL**: The gene name. This is *NEB* for all variants in this file.

We can easily create a new data frame with only those columns:

```{r}
df_anno$EFFECT <- gsub("&", ",", df_anno$EFFECT, fixed = TRUE) #Put in VEP format

df_vep_anno <- data.frame(PICK = rep(1, nrow(df_anno)),
                          Uploaded_variation = paste(2, df_anno$POS, df_anno$REF, df_anno$ALT, sep = ":"),
                          Consequence = df_anno$EFFECT,
                          LRT_pred = rep("D", nrow(df_anno)),
                          MutationTaster_pred = df_anno$MutationTaster_pred,
                          Polyphen2_HDIV_pred = df_anno$Polyphen2_HDIV_pred,
                          Polyphen2_HVAR_pred = df_anno$Polyphen2_HVAR_pred,
                          SIFT4G_pred = df_anno$SIFT_pred,
                          SYMBOL = df_anno$GENE)

head(df_vep_anno)
```

And as stated already, we will use all 3 annotation definitions for our masks and
we are using the *NEB* gene: 

```{r}
pLOF <- pLOF_narrowMissense <- pLOF_broadMissense <- TRUE
gene <- "NEB"
```

Now that we have put together all of the required parameters, we can call *agg_test()*:

```{r}
agg_test(score_stat_file, vcf_file, chr, burden, wburden,
                     SKAT, anno_file = NULL,
                     anno = df_vep_anno, two_stage = two_stage, 
                     two_stage_threshold = two_stage_threshold,
                     group_file = NULL, pLOF = pLOF,
                     pLOF_narrowMissense = pLOF_narrowMissense,
                     pLOF_broadMissense = pLOF_broadMissense,
                     altGroupFilePath = NULL, altCovariancePath = NULL,
                     altRaremetalPath = NULL, altRaremetalName = NULL,
                     mafThreshold = 0.01, gene = "NEB",
                     gene_start = gene_start, gene_end = gene_end, 
                     hwe = 0.000001)
```

Note that since the function returned a 1, at least one of the mask files passed the 
two-stage threshold, and RAREMETAL was called for each of the tests and masks. 
The pLOF results should hence be available at pLOF.meta.burden.results; the pLOF + missense(narrow)
results should be available at pLOF_narrowMissense.meta.burden.results; the 
pLOF + missense(broad) results should be available at pLOF_broadMissense.meta.burden.results

```{r}
pLOF_res <- fread("pLOF.meta.burden.results", header = TRUE, data.table = FALSE)
print(pLOF_res)
```


```{r}
pLOF_narrowMissense_res <- fread("pLOF_narrowMissense.meta.burden.results", header = TRUE, data.table = FALSE)
print(pLOF_narrowMissense_res)
```


```{r}
pLOF_broadMissense_res <- fread("pLOF_broadMissense.meta.burden.results", header = TRUE, data.table = FALSE)
print(pLOF_broadMissense_res)
```

Note we can also access the weighted burden results by changing the 'burden' 
to 'BBeta' in the file name, and the SKAT results by changing the 'burden' to 
'SKAT_'.

```{r}
pLOF_broadMissense_res_BBeta <- fread("pLOF_broadMissense.meta.BBeta.results", header = TRUE, data.table = FALSE)
print(pLOF_broadMissense_res_BBeta)
```

```{r}
pLOF_broadMissense_res_SKAT <- fread("pLOF_broadMissense.meta.SKAT_.results", header = TRUE, data.table = FALSE)
print(pLOF_broadMissense_res_SKAT)
```

Note that we are also able to supply the actual VCF used to generate the 
single-variant results. We would hope that the results using our external reference
panel above would be similar to our results using the African reference panel. 
This time we set two_stage to FALSE, and add a prefix to the RAREMETAL output 
via specifying altRaremetalPath = "trueCovariance_".

```{r}
vcf_file <- "vignetteData/chr2.1000G.european_unrel.NEB.vcf.gz"
agg_test(score_stat_file, vcf_file, chr, burden, wburden,
                     SKAT, anno_file = NULL,
                     anno = df_vep_anno, two_stage = FALSE, 
                     two_stage_threshold = two_stage_threshold,
                     group_file = NULL, pLOF = pLOF,
                     pLOF_narrowMissense = pLOF_narrowMissense,
                     pLOF_broadMissense = pLOF_broadMissense,
                     altGroupFilePath = NULL, altCovariancePath = NULL,
                     altRaremetalPath = "trueCovariance_", 
                     altRaremetalName = NULL, mafThreshold = 0.01, 
                     gene = "NEB", gene_start = gene_start, gene_end = gene_end, 
                     hwe = 0.000001)
```

The pLOF results should now be available at trueCovariance_pLOF.meta.burden.results; the pLOF + missense(narrow)
results should be available at trueCovariance_pLOF_narrowMissense.meta.burden.results; the 
pLOF + missense(broad) results should be available at trueCovariance_pLOF_broadMissense.meta.burden.results


```{r}
pLOF_res_insample <- fread("trueCovariance_pLOF.meta.burden.results", 
                           header = TRUE, data.table = FALSE)
print(pLOF_res_insample)
```


```{r}
pLOF_narrowMissense_res_insample <- fread("trueCovariance_pLOF_narrowMissense.meta.burden.results", 
                                          header = TRUE, data.table = FALSE)
print(pLOF_narrowMissense_res_insample)
```


```{r}
pLOF_broadMissense_res_insample <- fread("trueCovariance_pLOF_broadMissense.meta.burden.results", 
                                         header = TRUE, data.table = FALSE)
print(pLOF_broadMissense_res_insample)
```

As expected, the results using the African reference panel compared to the in-sample
covariance are quite similar for the pLOF and pLOF + missense(narrow) masks. The 
pLOF -log10(p-values) are `r -log10(pLOF_res$PVALUE)` and `r -log10(pLOF_res_insample$PVALUE)`
respectively. The pLOF + missense(narrow) -log10(p-values) are `r -log10(pLOF_narrowMissense_res$PVALUE)` 
and `r -log10(pLOF_narrowMissense_res_insample$PVALUE)` respectively, both of 
which are close to each other. The pLOF + missense(broad) -log10(p-values) are 
`r -log10(pLOF_broadMissense_res$PVALUE)` and `r -log10(pLOF_broadMissense_res_insample$PVALUE)` 
respectively, which are further apart, with the -log10(p-value) from using the
African reference panel being larger. This is not unexpected, as the reference panel
is both mismatched and small (<1000). We saw more discordance between -log10(p-values)
using InPSYght African American reference panels of size <1000 and in-sample UKBB 
covariance in our paper. This disconcordance is generally in the direction of the
reference panel giving a more significant -log10(p-value), as the reference panel 
does not capture as many of the variants and hence the majority of the covariance 
is being set to 0. 

## Performing Rare-Variant Aggregation Testing Given Group Files

A second option to using *agg_test()* is to supply a vector of strings containing
the file names of already created group files. Since *agg_test()* needs to write 
the group files before calling RAREMETAL, we can use the three masks created above
to show this. Let us use the African reference panel again, and skip using the 
two-stage approach. We set pLOF, pLOF_narrowMissense, and pLOF_broadMissense all 
to FALSE. If desired, they could be left to TRUE, and RAREMETAL will be run for
those 3 mask definitions as well as the user-supplied group files (which in this 
case are the same, and hence each test would be performed twice).

```{r}
vcf_file <- "vignetteData/chr2.1000G.african_unrel.NEB.vcf.gz"

group_file <- c("NEB.pLOF.group.file", 
               "NEB.pLOF_narrowMissense.group.file",
               "NEB.pLOF_broadMissense.group.file")

altRaremetalPath <- "user_supplied_groupFile_"

agg_test(score_stat_file, vcf_file, chr, burden, wburden,
                     SKAT, anno_file = NULL,
                     anno = df_vep_anno, two_stage = FALSE, 
                     two_stage_threshold = two_stage_threshold,
                     group_file = group_file, pLOF = FALSE,
                     pLOF_narrowMissense = FALSE,
                     pLOF_broadMissense = FALSE,
                     altGroupFilePath = NULL, altCovariancePath = NULL,
                     altRaremetalPath = altRaremetalPath, altRaremetalName = NULL,
                     mafThreshold = 0.01, gene = "NEB",
                     gene_start = gene_start, gene_end = gene_end, 
                     hwe = 0.000001)
```

In this case pLOF results will be at user_supplied_groupFile_userMask1.meta.burden.results;
pLOF + missense(narrow) results will be at user_supplied_groupFile_userMask2.meta.burden.results;
and pLOF + missense(broad) results will be auser_supplied_groupFile_userMask3.meta.burden.results

```{r}
pLOF_res_userMask <- fread("user_supplied_groupFile_userMask1.meta.burden.results", 
                           header = TRUE, data.table = FALSE)
print(pLOF_res_userMask)
```


```{r}
pLOF_narrowMissense_res_userMask <- fread("user_supplied_groupFile_userMask2.meta.burden.results", 
                                          header = TRUE, data.table = FALSE)
print(pLOF_narrowMissense_res_userMask)
```


```{r}
pLOF_broadMissense_res_userMask <- fread("user_supplied_groupFile_userMask3.meta.burden.results", 
                                         header = TRUE, data.table = FALSE)
print(pLOF_broadMissense_res_userMask)
```

These are the same as the results when not supplying the group files but setting
pLOF, pLOF_missenseNarrow, and pLOF_missenseBroad to TRUE. 

